### PR TITLE
<CheckToggle/> - new component

### PR DESCRIPTION
**Pull request by EgoziE**
Avishai's component, review it carefully 😬

### CODE REVIEW

**Comment by mosheKerbel** [comment]
@EgoziE 
Didn't really review it, just something that popped up
the `handleClick` is considered public in API docs, because it doesn't have `_` prefix (`_handleClick`)
is this the desired behaviour?

File: package.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
     "test:e2e": "npm run test:e2e-runtime && build-storybook -c ./.storybook/e2e && npm run test:e2e-only",
     "test:e2e-only": "NODE_ENV=production yoshi test --protractor",
     "test:e2e-runtime": "cd ./test-runtime/imports && NODE_ENV=production npm run test",
+    "pretest:visual": "npm run generate-autodocs-registry",
```
Comment (shlomitc) [inline package.json:47]: why is this needed?

File: src/CheckToggle/CheckToggle.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+   * Except when disabled
+   */
+  handleClick = () => {
+    const { checked } = this.state;
+    const { disabled, onChange } = this.props;
+
+    if (disabled) return;
+
+    this.setState({ checked: !checked });
+    if (onChange) onChange(!checked);
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:33]: have this as a callback function to `setState` to ensure it happens after the render

File: src/CheckToggle/CheckToggle.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * Renders the button itself
+   * @returns React.ReactNode
+   * @private
+   */
+  _renderButton = () => {
+    const { checked } = this.state;
+    const { size, skin, disabled } = this.props;
+    return (
+      <button
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:45]: doesn't it make sense to use `input type="checkbox"` instead of a `button`?
this would also make it possible to use the component in an `uncontrolled` way

File: package.json (hunk 1 above)
Comment (EgoziE) [reply package.json:47]: I'm using autodocs to get all components metadata to create permutations

Comment (shlomitc) [reply package.json:47]: Then you will need this also in `storybook:visual`.
Please note that this action adds 16s to the build step.
In general, I think that this task should be some **storybook addon**, so it will be configured part of the storybook and not part of the build step - @argshook WDYT?

File: src/CheckToggle/CheckToggle.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    if (disabled) return;
+
+    this.setState({ checked: !checked }, () => {
+      if (onChange) onChange(!checked);
+    });
+  };
+
+  /**
+   * Renders the button itself
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:38]: a "toggle", not a button

File: src/CheckToggle/CheckToggle.st.css [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+.input {
+  position: absolute;
+  opacity: 0;
+  cursor: pointer;
+  height: 0;
+  width: 0;
+}
+
+.button {
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.st.css:36]: if it's no longer a button, but a toggle, let's change the terminology.

File: src/CheckToggle/CheckToggle.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+   * Toggles checked state and triggers the onChange callback.
+   * Except when disabled
+   */
+  _handleClick = () => {
+    const { checked } = this.state;
+    const { disabled, onChange } = this.props;
+
+    if (disabled) return;
+
+    this.setState({ checked: !checked }, () => {
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:32]: The existing beahvior mixes the controlled and uncontrolled pattern. 
I'd expect `<CheckToggle skin="standard" checked />` to not change the checked state when I click on it.

Take the native `input`:
`<input/>` - this will behave as an `uncontrolled`, meaning it will call `onChange` for every character type and update the internal state value.
`<input value=""/>` - this will behave as `controlled` and will call the `onChange`, but will **not** change the internal state unless explicitly passed as props.

so I think you should also check in this component if you're in controlled mode or not

File: src/CheckToggle/CheckToggle.st.css [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  -st-extends: Focusable;
+  -st-states: checked, disabled, size(enum(small, medium)),
+    skin(enum(standard, success)), size(enum(small, medium));
+
+  outline: none;
+  border-radius: 100%;
+}
+
+/* Input */
+.label {
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.st.css:21]: any reason why we need an extra `label` and not merge it with the `root` div?

File: src/CheckToggle/docs/examples.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
+export const skin = `
+<Container>
```
Comment (shlomitc) [inline src/CheckToggle/docs/examples.js:2]: I think you mixed a bit the `Grid` component with the `Layout` one.

If you use `Col`, you should have `Row` before. `Container` sets min-max width which isn't necessary here.

```jsx
<Layout>
  <Cell span={3}>
    <CheckToggle skin="standard" checked />
  </Cell>
  <Cell span={9}>
    <CheckToggle skin="standard" />
  </Cell>
  <Cell span={3}>
    <CheckToggle skin="success" checked />
  </Cell>
  <Cell span={9}>
    <CheckToggle skin="success" />
  </Cell>
</Layout>
```

File: src/CheckToggle/docs/examples.js [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  <Col span={1}>
+    <CheckToggle skin="success" checked />
+  </Col>
+  <Col span={1}>
+    <CheckToggle skin="success" />
+  </Col>
+</Container>
+`;
+
+export const size = `
```
Comment (shlomitc) [inline src/CheckToggle/docs/examples.js:24]:
```jsx
<Layout>
  <Cell span={3}>
    <CheckToggle checked size="medium"/>
  </Cell>
  <Cell span={9}>
    <CheckToggle  size="medium"/>
  </Cell>
  <Cell span={3}>
    <CheckToggle checked  size="small"/>
  </Cell>
  <Cell span={9}>
    <CheckToggle  size="small"/>
  </Cell>
</Layout>
```

File: src/CheckToggle/docs/examples.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  <Col span={1}>
+    <CheckToggle size="medium" checked />
+  </Col>
+  <Col span={1}>
+    <CheckToggle size="medium" />
+  </Col>
+</Container>
+`;
+
+export const disabled = `
```
Comment (shlomitc) [inline src/CheckToggle/docs/examples.js:47]: like the above

File: src/CheckToggle/docs/examples.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  <Col span={1}>
+    <CheckToggle disabled checked />
+  </Col>
+  <Col span={1}>
+    <CheckToggle disabled />
+  </Col>
+</Container>
+`;
+
+export const tooltip = `
```
Comment (shlomitc) [inline src/CheckToggle/docs/examples.js:58]: like the above

File: src/CheckToggle/CheckToggle.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+  /**
+   * Renders the toggle itself
+   * @returns React.ReactNode
+   * @private
+   */
+  _renderInput = () => {
+    const { size, disabled, onChange } = this.props;
+    let _onChange = null;
+    if (!disabled) {
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:54]: Since you're rendering input that already handles all input props, why not pass the `disable` prop as well?

I find it a bit confusing that you're conditioning it on the `disabled` while you can actually condition this on the `isControlled`.

as I see it
```js
if(controlled) {
 inputProps = {onChange: this._handleChange, value: props.value, disabled: props.disabled}
} else {
 inputProps = {onChange: props.onChange, disabled: props.disabled}
}
```

File: src/CheckToggle/CheckToggle.js [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+      disabled,
+      tooltipContent,
+      focusableOnFocus,
+      focusableOnBlur,
+    } = this.props;
+
+    return (
+      <label
+        {...styles('root', { checked, size, skin, disabled }, this.props)}
+        data-checked={checked}
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.js:104]: you can just check the actual input instead of adding a custom data- attribute

File: src/CheckToggle/CheckToggle.uni.driver.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    /** Get the current count */
+    isChecked: async () =>
+      (await base.$(`[data-hook="${dataHooks.toggle}"]`).getNative()).checked, // eslint-disable-line no-restricted-properties
+
+    /** Click the check toggle */
+    click: async () => {
+      const input = await base.$(`[data-hook="${dataHooks.toggle}"]`);
+
+      // eslint-disable-next-line no-restricted-properties
+      const isDisabled = (await input.getNative()).disabled;
```
Comment (shlomitc) [inline src/CheckToggle/CheckToggle.uni.driver.js:21]: Using `getNative()` is not allowed, as it breaks the whole concept of Unidriver.
I'd want this to be natively in Unidriver or use ReactBase as a last resort

Comment (mosheKerbel) [reply src/CheckToggle/CheckToggle.uni.driver.js]: You have the `_prop` function in UniDriver
So you can 
`const isDisabled = (await input._prop('disabled'))`
